cupyx/scipy/sparse/construct.py

```diff
@@ -88,6 +88,230 @@ def spdiags(data, diags, m, n, format=None):
+        dtype (dtype, optional):
+            The data-type of the output matrix.  If not given, the dtype is
+            determined from that of ``blocks``.
+
+    Returns:
+        cupyx.scipy.sparse.spmatrix: the stacked sparse matrix
+
+    .. seealso:: :func:`scipy.sparse.hstack`
+
+    Examples:
+        >>> from cupy.sparse import coo_matrix, hstack
+        >>> A = coo_matrix([[1, 2], [3, 4]])
+        >>> B = coo_matrix([[5], [6]])
+        >>> hstack([A,B]).toarray()
+        array([[1, 2, 5],
+               [3, 4, 6]])
+    """
+    return bmat([blocks], format=format, dtype=dtype)
+
+
+def vstack(blocks, format=None, dtype=None):
+    """Stacks sparse matrices vertically (row wise)
+    Args:
+        blocks (sequence of cupyx.scipy.sparse.spmatrix)
+            sparse matrices to stack
+        format (str, optional):
+            sparse format of the result (e.g. "csr")
+            by default an appropriate sparse matrix format is returned.
+            This choice is subject to change.
+        dtype (dtype, optional):
+            The data-type of the output matrix.  If not given, the dtype is
+            determined from that of `blocks`.
+
+    Returns:
+        cupyx.scipy.sparse.spmatrix: the stacked sparse matrix
+
+    .. seealso:: :func:`scipy.sparse.vstack`
+
+    Examples:
+        >>> from cupy.sparse import coo_matrix, vstack
+        >>> A = coo_matrix([[1, 2], [3, 4]])
+        >>> B = coo_matrix([[5, 6]])
+        >>> vstack([A, B]).toarray()
+        array([[1, 2],
+               [3, 4],
+               [5, 6]])
+    """
+    return bmat([[b] for b in blocks], format=format, dtype=dtype)
+
+
+def bmat(blocks, format=None, dtype=None):
+    """Builds a sparse matrix from sparse sub-blocks
+    Args:
+        blocks (array_like):
+            Grid of sparse matrices with compatible shapes.
+            An entry of None implies an all-zero matrix.
+        format ({'bsr', 'coo', 'csc', 'csr', 'dia', 'dok', 'lil'}, optional):
+            The sparse format of the result (e.g. "csr").  By default an
+            appropriate sparse matrix format is returned.
+            This choice is subject to change.
+        dtype (dtype, optional):
+            The data-type of the output matrix.  If not given, the dtype is
+            determined from that of `blocks`.
+    Returns:
+        bmat (sparse matrix)
+
+    .. seealso:: :func: `scipy.sparse.bmat`
+
+    Examples:
+        >>> from cupy.sparse import coo_matrix, bmat
+        >>> A = coo_matrix([[1, 2], [3, 4]])
+        >>> B = coo_matrix([[5], [6]])
+        >>> C = coo_matrix([[7]])
+        >>> bmat([[A, B], [None, C]]).toarray()
+        array([[1, 2, 5],
+               [3, 4, 6],
+               [0, 0, 7]])
+        >>> bmat([[A, None], [None, C]]).toarray()
+        array([[1, 2, 0],
+               [3, 4, 0],
+               [0, 0, 7]])
+    """
+
+    # We assume here that blocks will be 2-D so we need to look, at most,
+    # 2 layers deep for the shape
```

Comment:
Let's check this assumption and raise `ValueError` if violated.
TODO is also OK.